Added card assets endpoint in the admin

minimaluminium opened the pull request:
ref https://linear.app/ghost/issue/PROD-2481/oss-issue-cardsmincss-and-js-problems-for-self-hosters

- Added `/ghost/cards/` route for theme-independent CSS/JS card assets
- Updated ActivityPub reader to use admin card assets instead of theme-dependent ones
- Ensured all card styles and functionality available regardless of theme settings

minimaluminium committed:
Added card assets endpoint in the admin

minimaluminium committed:
Used relative path for the assets

kevinansfield commented on apps/admin-x-activitypub/src/views/Inbox/components/Reader.tsx:
```diff
@@ -154,12 +151,12 @@ const ArticleBody: React.FC<{
 
                     const script = document.createElement('script');
-                    script.src = '${siteUrl ? `${siteUrl.replace(/\/$/, '')}/public/cards.min.js` : '/public/cards.min.js'}';
+                    script.src = 'cards/admin-cards.min.js';
```
This stuck out initially because it makes the script URL dependent on the currently-viewed page, _but_ I think this will be fine as long as we're using `#` URLs for the SPA

minimaluminium replied:
Switched to absolute URLs in https://github.com/TryGhost/Ghost/pull/24746/commits/3d0d206df1bef0b296f3919d809ab6a74600bf39

kevinansfield reviewed (changes requested):
Code looks good barring the potential issue with multiple simultaneous builds but we're completely missing tests for the new admin app endpoint. It would be great to see the tests added along with some tests for the "still running/ready" behaviour.

minimaluminium committed:
Use absolute URL for the card assets in the Reader

minimaluminium committed:
Added protection against multiple request making a conflict

Prevent multiple simultaneous builds by implementing promise caching pattern.
Multiple requests during initial load could cause file write conflicts when
trying to build assets concurrently.

minimaluminium committed:
Add tests for admin card assets endpoint

Add missing test coverage for the new admin card assets endpoint and
concurrent loading protection as requested in code review.

minimaluminium commented on ghost/core/core/server/web/admin/card-assets.js:
```diff
@@ -0,0 +1,168 @@
+        if (!this.ready) {
+            await this.load();
+        }
+    }
```
Added a multiple connection protection in https://github.com/TryGhost/Ghost/pull/24746/commits/60a8690e3f9ed772c64defcc3b1de69c6d22fb5e

minimaluminium committed:
Await async fixture setup/cleanup to avoid racey tests

minimaluminium committed:
Fix should.js assertions missing parentheses

minimaluminium committed:
Use fs.rm instead of deprecated fs.rmdir

minimaluminium committed:
Fixed 404 error test for non existent file

minimaluminium committed:
Reverted back to the subdir compatible URLs
